_posts/2021-11-23-how-to-dry-your-ref-arch.adoc

```diff
@@ -0,0 +1,606 @@
+Service Catalog Terraform module takes in the container image in the format:
+
+[source,hcl]
+----
+type = object({
+  # Repository of the docker image (e.g. gruntwork/frontend-service)
+  repository = string
+  # The tag of the docker image to deploy.
+  tag = string
+  # The image pull policy. Can be one of IfNotPresent, Always, or Never.
+  pull_policy = string
+})
+----
+
+It is typical for only the `tag` attribute to be different across environments. In the guide, you would have to repeat
+this configuration across the environments, but with deep merge, you can promote the `repository` and `pull_policy`
+attributes to the common component configuration:
+
+.common component configuration
+[source,hcl]
+----
+inputs = {
+  container_image = {
+    repository  = "gruntwork/aws-sample-app"
+    pull_policy = "IfNotPresent"
+  }
+}
+----
+
+.dev configuration
+[source,hcl]
+----
+include "parent" {
+  path           = "/path/to/parent/configuration"
+  merge_strategy = "deep"
+}
+
+inputs = {
+  container_image = {
+    tag = "v0.0.4"
+  }
+}
+----
+
+.stage configuration
+[source,hcl]
+----
+include "parent" {
+  path           = "/path/to/parent/configuration"
+  merge_strategy = "deep"
+}
+
+inputs = {
+  container_image = {
+    tag = "v0.0.3"
+  }
+}
+----
+
+In this way, you can leverage deep merge to refactor complex nested inputs in your Terragrunt configuration to further
+DRY up the config.
+
+
+[[expose_include]]
+===== Expose `include`
+
+Terragrunt offers the ability to reference values defined in parent configurations through exposed includes. This
+feature is enabled when the `expose = true` attribute is set on the `include` block. Exposed `include` blocks allow the
+child configuration to reference values that are defined in the parent configuration. These values are available with
+the reference `include.LABEL`.
+
+This is most useful to reference `locals` that are defined in the common component configuration to reuse them in the
+child.
+
+For example, if you wanted to test a new version of the module only in dev:
+
+.common component configuration
+[source,hcl]
+----
+locals {
+  source_base_url = "git::git@github.com:gruntwork-io/terraform-aws-service-catalog.git//modules/networking/vpc"
+}
+
+terraform {
+  source = "${local.source_base_url}?ref=v0.65.0"
+}
+----
+
+.dev configuration
+[source,hcl]
+----
+include "parent" {
+  path   = "/path/to/parent/configuration"
+  expose = true
+}
+
+terraform {
+  source = "${include.parent.locals.source_base_url}?ref=v0.66.0"
+}
+----
+
+Note that the availability of values is subject to the
+https://terragrunt.gruntwork.io/docs/etting-started/configuration/#configuration-parsing-order[configuration parsing
+order] of Terragrunt.
+This means that you won't be able to reference later stage values in early stage blocks, like accessing parent `inputs`
+in `locals` or trying to reference a `dependency` in `locals`.
```

Comment:
So, the website is down at the moment and I forgot the config parsing order, but I can intuit that `inputs` come last. I'm guessing it goes, `locals`, then `dependency`, then `inputs`, because it's possible to reference locals within dependency blocks.

I think a little more wording would make it just a bit clearer:

```suggestion
This means that you won't be able to reference later stage values in early stage blocks, like accessing parent `inputs`
in child `locals` or trying to reference a parent `dependency` in child `locals`.
```

Reply:
Good point! Done in https://github.com/gruntwork-io/gruntwork-io.github.io/pull/669